### PR TITLE
Support azure signtool

### LINKED ISSUE #767: Support AzureSignTool
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

[AzureSignTool](https://github.com/vcsjones/AzureSignTool/) is another tool that can be used to sign binaries. It uses a vault to obtain the certificate instead of using a certificate file.

`constructor` should extend its support to different signing tools and also support key-vault signature processes.

### Why is this needed?

Using vaults to sign binaries is going to be more and more important. 
`AzureSignTool` uses a vault without files, so signing installers is not possible with this

As of now, `constructor` uses `signtool.exe` and local files to sign binaries, so `AzureSignTool` or other file-less signature tools will not work.

### What should happen?

* Expanding `construct.yaml` to allow for vaults to sign packages.
* Allow for other tools to sign binaries, starting with `AzureSignTool`.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by marcoesters**
### Description

Add support for AzureSignTool to sign Windows installers. `AzureSignTool` uses Azure key vaults, which do not use certificate files on disk. To achieve this, this PR:

* Adds a `windows_signing_tool` keyword that selects the tool to sign. This key determines whether to sign the installer instead of `signing_certificate`, which makes signing more extensible.
* Creates classes for Windows signing tools.
* Adds documentation for all environment variables for both `SignTool` and `AzureSignTool`.
* Allows `signtool.exe` to use a different digest algorithms.

Since this requires access to Azure, tests cannot be added for this feature.

Closes #767.

### Checklist - did you ...

- [X] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
- [X] Add / update necessary tests?
- [X] Add / update outdated documentation?

### CODE REVIEW

File: docs/source/howto.md
Comment (jaimergp) [inline docs/source/howto.md:89]:
```suggestion
:::
```

File: docs/source/howto.md
Comment (jaimergp) [inline docs/source/howto.md:90]:
```suggestion
### Signing and notarizing PKG installers
```

File: news/771-add-azure-signtool-support
Comment (jaimergp) [inline news/771-add-azure-signtool-support:3]:
```suggestion
* Add support for AzureSignTool to sign Windows installers. (#767 via #771)
```

File: constructor/construct.py
Comment (jaimergp) [inline constructor/construct.py:806]: This reminds me we need to start raising exceptions instead of sys.exits, but that's for another PR.

File: constructor/main.py
Comment (jaimergp) [reply constructor/main.py:114]: At the same time, if we are validating the input in construct.py:805, why do we need normalization here? 🤔 For library usage? In that case it should be the same logic (or move the logic here).

File: constructor/main.py
Comment (jaimergp) [inline constructor/main.py:116]: We should check if we are overriding something here, and warn appropriately.

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:93]:
```suggestion
        super()._verify_tool_is_available()
```

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:128]: I think this can be easily moved to the base class as its own (static?) method.

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:159]: Can we `log.info` which method was detected?

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:185]:
```suggestion
        https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.signaturestatus
```

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:207]: Should we raise here? 🤔

File: constructor/winexe.py
Comment (jaimergp) [inline constructor/winexe.py:213]: Can we add some typing? There are a few things called `signtool` or similar now and it can get confusing. I think this expects an instance of the new class in this PR?

**Review by jaimergp — COMMENTED** [review]
Very nice work. I'm a little concerned about the lack of tests, so I wonder if it's feasible to have a some secrets added to a dummy vault (or something that implements the same API). However, that might be too complicated and this feature will be used routinely by Anaconda anyway, so I guess we'll know when it breaks.

Double check that the docs render correctly locally, I found a little typo with the admonition syntax. Other than that, this is good to go after addressing the comments.

File: constructor/main.py
Comment (marcoesters) [reply constructor/main.py:114]: > What happens if a user tries to set this to a full path? 

`constructor` will and should fail. `windows_signing_tool` is the name. Otherwise, I would have to infer from the binary what the appropriate signing too is.

>  In that case it should be the same logic

Yes, that's true, good catch!

> (or move the logic here).

I'm not sure about this. Setting default values and manipulating input is done in `main_build`. `validate()` performs validation here. If we move this line into `validate()`, we are mixing scopes, which makes things harder to track.

File: constructor/main.py
Comment (marcoesters) [reply constructor/main.py:116]: That check is done in line 113 - if `windows_signing_tool` is set in any way, we won't get to `elif`.

File: constructor/signing.py
Comment (marcoesters) [reply constructor/signing.py]: Since this is a very general helper function, I moved it to utils

**Comment by marcoesters** [comment]
I added integration tests and secrets were added to the repository. However, the secrets will not be available for anybody who submits a PR from the fork, which will cause the tests to fail. The tests do execute the correct command though: https://github.com/conda/constructor/actions/runs/8760110611/job/24044501617#step:9:2133

Those tests will have to be skipped until merged into `main`. So, we will have to add 
As discussed with @jaimergp offline, the best way to go here is to add test failure reporting as they do for `conda-libmambda-solver`: https://github.com/conda/conda-libmamba-solver/blob/efa3b84141e4ff777928aaae952fac417b059091/.github/workflows/tests.yml#L680-L690

This requires a token, so I will convert this PR into draft until that infrastructure is in place.

Additional things I found:

* The environment is not copied into the subprocess on Windows when no extra environment variables are set. On POSIX, this is handled by the subprocess module, but not for Windows. So, I copy the environment explicitly for all to get conistent behavior.
* The debugging section was not in any TOC tree for the documentations, so I added it.

File: .github/workflows/main.yml
Comment (jaimergp) [inline .github/workflows/main.yml:98]:
```suggestion
        if: matrix.os == 'windows'
```
These can be omitted in `if` clauses, I think.

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:85]: Just noting here that no quotes are needed because we don't expect spaces here.

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:145]: Should we quote this?

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:168]: Same, do we need to quote any of these?

File: constructor/signing.py
Comment (jaimergp) [inline constructor/signing.py:208]: Can this fail if status is empty or None?

File: tests/test_examples.py
Comment (jaimergp) [inline tests/test_examples.py:493]: Why do we need this "parametrized" parametrization? Are you using it locally while debugging? I don't see it in the CI setup 🤔

File: constructor/signing.py
Comment (marcoesters) [reply constructor/signing.py:208]: I'm not sure how this could be `None` unless `proc.stdout` can somehow be `None`.

If `status` is and empty string, `int(status)` raises a `ValueError`, which is caught further down.

File: tests/test_examples.py
Comment (marcoesters) [reply tests/test_examples.py:493]: I wanted to make sure that all authentication methods could potentially be covered. For tokens and secrets, I can just check the environment variables and skip if needed. For managed identities, however, this does not work because there isn't an environment variable to check for (logging into the vault must happen before AzureSignTool is executed).

So, the only "safe" methods to add into the parametrization are `token` and `secret`. However, I still wanted to allow testing with a managed identity.

Does that make sense? I tried to explain this briefly in the docstring.

File: constructor/signing.py
Comment (marcoesters) [reply constructor/signing.py]: We can. I just tested it locally and the test worked fine with quotes, too.

File: tests/test_examples.py
Comment (jaimergp) [inline tests/test_examples.py:499]:
```suggestion
    There is no good sentinel environment for managed identities, so an environment variable
```

I think?

File: docs/source/howto.md
Comment (jaimergp) [inline docs/source/howto.md:90]:
```suggestion

### Signing and notarizing PKG installers
```

**Review by jaimergp — APPROVED** [review]

**Pull request merged**
